flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobResourceRequirements.java

```diff
@@ -0,0 +1,204 @@
+            final Optional<Integer> maybeMaxParallelism =
+                    Optional.ofNullable(maxParallelismPerVertex.get(jobVertexId));
+            if (maybeMaxParallelism.isPresent()) {
+                final JobVertexResourceRequirements.Parallelism requestedParallelism =
+                        jobResourceRequirements.findParallelism(jobVertexId).get();
+                if (requestedParallelism.getLowerBound() > requestedParallelism.getUpperBound()) {
+                    errors.add(
+                            String.format(
+                                    "The requested lower bound [%s] for job vertex [%s] is higher than the upper bound [%d].",
+                                    requestedParallelism.getLowerBound(),
+                                    jobVertexId,
+                                    requestedParallelism.getUpperBound()));
+                }
+                if (maybeMaxParallelism.get() < requestedParallelism.getUpperBound()) {
+                    errors.add(
+                            String.format(
+                                    "The newly requested parallelism %d for the job vertex %s exceeds its maximum parallelism %d.",
+                                    requestedParallelism.getUpperBound(),
+                                    jobVertexId,
+                                    maybeMaxParallelism.get()));
+                }
+            } else {
+                errors.add(
+                        String.format(
+                                "Job vertex [%s] was not found in the JobGraph.", jobVertexId));
+            }
+            missingJobVertexIds.remove(jobVertexId);
+        }
+        for (JobVertexID jobVertexId : missingJobVertexIds) {
+            errors.add(
+                    String.format(
+                            "The request is incomplete, missing job vertex [%s] resource requirements.",
+                            jobVertexId));
+        }
+        return errors;
+    }
+
+    public static JobResourceRequirements empty() {
+        return JobResourceRequirements.EMPTY;
+    }
+
+    public static Builder newBuilder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+
+        private final Map<JobVertexID, JobVertexResourceRequirements> vertexResources =
+                new HashMap<>();
+
+        public Builder setParallelismForJobVertex(
+                JobVertexID jobVertexId, int lowerBound, int upperBound) {
+            vertexResources.put(
+                    jobVertexId,
+                    new JobVertexResourceRequirements(
+                            new JobVertexResourceRequirements.Parallelism(lowerBound, upperBound)));
+            return this;
+        }
+
+        public JobResourceRequirements build() {
+            return new JobResourceRequirements(
+                    Collections.unmodifiableMap(new HashMap<>(vertexResources)));
```

Comment:
Should the defensive copying be performed in the constructor?

Reply:
Good point 👍 initially the constructor was private, and the builder was the only way of constructing this, but it had changed with some recent changes